### PR TITLE
eventqueue: protect against enqueueing same Event twice

### LINKED ISSUE #8805: panic in EventQueue occurs when multiple calls to Enqueue occur for same Event
simulating some scheduling by introducing a time.Sleep it seems it is possible the eventqueue to trigger a panic while running.
```golang
diff --git a/pkg/eventqueue/eventqueue.go b/pkg/eventqueue/eventqueue.go
index de1f8af86..87823d4d2 100644
--- a/pkg/eventqueue/eventqueue.go
+++ b/pkg/eventqueue/eventqueue.go
@@ -17,11 +17,13 @@ package eventqueue
 import (
 	"reflect"
 	"sync"
+	"time"
 
 	"github.com/cilium/cilium/pkg/logging"
 	"github.com/cilium/cilium/pkg/logging/logfields"
 	"github.com/cilium/cilium/pkg/option"
 	"github.com/cilium/cilium/pkg/spanstat"
+
 	"github.com/sirupsen/logrus"
 )
 
@@ -170,7 +172,7 @@ func (ev *Event) WasCancelled() bool {
 // stopped, the Event will not be enqueued, and its cancel channel will be
 // closed, indicating that the Event was not ran. This function may block if
 // the queue is at its capacity for events.
-func (q *EventQueue) Enqueue(ev *Event) <-chan interface{} {
+func (q *EventQueue) Enqueue(ev *Event, d time.Duration) <-chan interface{} {
 
 	if q.notSafeToAccess() || ev == nil {
 		return nil
@@ -178,6 +180,8 @@ func (q *EventQueue) Enqueue(ev *Event) <-chan interface{} {
 
 	// Track that event has been Enqueued.
 	q.closeWaitGroup.Add(1)
+	// simulate scheduling
+	time.Sleep(d)
 	defer q.closeWaitGroup.Done()
 
 	select {
diff --git a/pkg/eventqueue/eventqueue_test.go b/pkg/eventqueue/eventqueue_test.go
index 433a59c34..a66d45480 100644
--- a/pkg/eventqueue/eventqueue_test.go
+++ b/pkg/eventqueue/eventqueue_test.go
@@ -84,7 +84,7 @@ func (s *EventQueueSuite) TestDrained(c *C) {
 
 func (s *EventQueueSuite) TestNilEvent(c *C) {
 	q := NewEventQueue()
-	res := q.Enqueue(nil)
+	res := q.Enqueue(nil, 0)
 	c.Assert(res, IsNil)
 }
 
@@ -105,14 +105,14 @@ func (s *EventQueueSuite) TestEventCancelAfterQueueClosed(c *C) {
 	q := NewEventQueue()
 	q.Run()
 	ev := NewEvent(&DummyEvent{})
-	q.Enqueue(ev)
+	q.Enqueue(ev, 0)
 
 	// Event should not have been cancelled since queue was not closed.
 	c.Assert(ev.WasCancelled(), Equals, false)
 	q.Stop()
 
 	ev = NewEvent(&DummyEvent{})
-	q.Enqueue(ev)
+	q.Enqueue(ev, 0)
 	c.Assert(ev.WasCancelled(), Equals, true)
 }
 
@@ -133,6 +133,21 @@ func CreateHangEvent() *NewHangEvent {
 	}
 }
 
+func (s *EventQueueSuite) TestPanic(c *C) {
+	q := NewEventQueue()
+	q.Run()
+
+	nh1 := CreateHangEvent()
+	ev := NewEvent(nh1)
+	go q.Enqueue(ev, time.Second)
+	// stop the queue but at the "same" time a new event is scheduled
+	q.Stop()
+	time.Sleep(500*time.Millisecond)
+	q.Enqueue(ev, time.Second)
+
+	q.WaitToBeDrained()
+}
+
 func (s *EventQueueSuite) TestDrain(c *C) {
 	q := NewEventQueue()
 	q.Run()
@@ -142,19 +157,19 @@ func (s *EventQueueSuite) TestDrain(c *C) {
 	nh3 := CreateHangEvent()
 
 	ev := NewEvent(nh1)
-	q.Enqueue(ev)
+	q.Enqueue(ev, 0)
 
 	ev2 := NewEvent(nh2)
 	ev3 := NewEvent(nh3)
 
-	q.Enqueue(ev2)
+	q.Enqueue(ev2, 0)
 
 	var rcvChan <-chan interface{}
 
 	enq := make(chan struct{})
 
 	go func() {
-		rcvChan = q.Enqueue(ev3)
+		rcvChan = q.Enqueue(ev3, 0)
 		enq <- struct{}{}
 	}()
 ```

### PR DESCRIPTION
**Pull request by ianvernon**
Previously, if an Event is enqueued twice, the following panic would occur:

```
PANIC: eventqueue_test.go:187: EventQueueSuite.TestPanic

... Panic: close of closed channel (PC=0x42EFD4)

/usr/local/go/src/runtime/panic.go:522
  in gopanic
/usr/local/go/src/runtime/chan.go:342
  in closechan
eventqueue.go:188
  in EventQueue.Enqueue
eventqueue_test.go:197
  in EventQueueSuite.TestPanic
/usr/local/go/src/reflect/value.go:308
  in Value.Call
/usr/local/go/src/runtime/asm_amd64.s:1337
  in goexit
OOPS: 9 passed, 1 PANICKED
--- FAIL: Test (1.51s)
FAIL
coverage: 0.0% of statements in pkg/eventqueue
FAIL	github.com/cilium/cilium/pkg/eventqueue	1.550s
Makefile:129: recipe for target 'unit-tests' failed
make: *** [unit-tests] Error 1
```

To protect against this, do not allow enqueueing the same Event twice, by using
an atomic boolean which contains information about whether a given Event has
been enqueued or not.

Fixes: #8805

Signed-off by: Ian Vernon <ian@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/8809)

### CODE REVIEW

**Comment by ianvernon** [comment]
test-me-please

File: pkg/eventqueue/eventqueue.go [hunk 1]
```diff
@@ -169,11 +176,16 @@ func (ev *Event) WasCancelled() bool {
...
-
 	if q.notSafeToAccess() || ev == nil {
 		return nil
 	}
 
+	// Events can only be enqueued once.
+	if val := ev.enqueued.Load(); val == true {
+		return nil
+	}
+	ev.enqueued.Store(true)
```
Comment (jrajahalme) [inline pkg/eventqueue/eventqueue.go]: This does not prevent two concurrent Enqueue()s for the same event, if they both Load `enqueued` before them both storing `true`. Maybe `if atomic.AddUint32(&ev.enqueued, 1) > 1 { return nil }` could be used instead?

**Review by jrajahalme — CHANGES_REQUESTED** [review]
Still a chance for a tight race.

File: pkg/eventqueue/eventqueue.go (hunk 1 above)
Comment (ianvernon) [reply pkg/eventqueue/eventqueue.go]: You are correct, done!

**Comment by ianvernon** [comment]
test-me-please

File: pkg/eventqueue/eventqueue.go [hunk 2]
```diff
@@ -142,12 +147,13 @@ type eventStatistics struct {
 
 // NewEvent returns an Event with all fields initialized.
 func NewEvent(meta EventHandler) *Event {
-	return &Event{
+	ev := Event{
```
Comment (aanm) [inline pkg/eventqueue/eventqueue.go]: unnecessary change

File: pkg/eventqueue/eventqueue.go [hunk 3]
```diff
@@ -169,11 +175,15 @@ func (ev *Event) WasCancelled() bool {
 // closed, indicating that the Event was not ran. This function may block if
 // the queue is at its capacity for events.
 func (q *EventQueue) Enqueue(ev *Event) <-chan interface{} {
-
 	if q.notSafeToAccess() || ev == nil {
 		return nil
 	}
 
+	// Events can only be enqueued once.
+	if atomic.AddInt32(&ev.enqueued, 1) > 1 {
```
Comment (aanm) [inline pkg/eventqueue/eventqueue.go:188]: ideally this should be a sync.Once, otherwise the event order is not being kept, i.e. if we enqueue the same event at the same time, the response of the 2nd queued might, and probably will, be returned before the 1st queued is executed.

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/eventqueue/eventqueue.go (hunk 3 above)
Comment (ianvernon) [reply pkg/eventqueue/eventqueue.go:188]: If we Enqueue the same event at the same time, one of the events will add to the int atomically, and the other one won't - how does that differ from a `sync.Once`? I can theoretically use a `sync.Once` since that is basically what I am using here (`sync.Once` is a wrapper around a mutex and `sync.Atomic` loading and storing a `uint32`.  

In other words, ordering will never be guaranteed in this case, it's a matter of who gets there first. I think that's fine. The `EventQueue` wasn't designed to have the same `Event` enqueued multiple times; safeguarding against that with some state which signifies whether it has been Enqueued is a needed fix, but I do not think that we need to ensure that if the same event is enqueued multiple times that we need to preserve the ordering of who calls Enqueue - that would mean that multiple goroutines would be Enqueueing the same event at the same time.

**Comment by ianvernon** [comment]
test-me-please

File: pkg/eventqueue/eventqueue.go (hunk 3 above)
Comment (aanm) [reply pkg/eventqueue/eventqueue.go:188]: >how does that differ from a sync.Once?

The sync.Once will wait until all other go routines being ran inside the Once are executed.

Comment (ianvernon) [reply pkg/eventqueue/eventqueue.go:188]: I do not understand what advantages of waiting for the goroutines being ran inside the `sync.Once` are executed provides us is in this case. The critical section here is quite small as opposed to wrapping the contents of `Enqueue` itself within a `sync.Once`. Can you elaborate on what advantages that provides in this case?

Comment (aanm) [reply pkg/eventqueue/eventqueue.go:188]: >Can you elaborate on what advantages that provides in this case?

Why is the function named "Enqueue" if we will return earlier in a 2nd, and following "enqueues" of an event that is still in the queue and still being processed? By earlier I meant that we will return even before the event was processed the first time.

Comment (ianvernon) [reply pkg/eventqueue/eventqueue.go:188]: The first enqueue will process the event, the second enqueue will return a nil result immediately, which the user can check for. A `nil` return means that the Enqueue was a no-op. I can write that into the documentation. When would the ordering of returns matter in this case? I do not ever see a case in which a user would enqueue the same event twice across different goroutines. With this PR, we guard against that by returning a nil value for the second call to Enqueue. Can you provide an example where the ordering would matter? I feel like this is a semantic argument as opposed to one of correctness.

Comment (aanm) [reply pkg/eventqueue/eventqueue.go:188]: It's weird to have a function called "enqueue" for which in reality is not enqueing events.

Comment (ianvernon) [reply pkg/eventqueue/eventqueue.go:188]: The code was designed such that a given event should be enqueued only once. If an event is enqueued more than once, we return a signal to the user (`nil` result) that it was not enqueued. I could return an error as well from `Enqueue` indicating as such? If a user wants to `Enqueue` similar information again, they can call `NewEvent` and `Enqueue` that information. The path you are talking about is not a use-case that was envisioned when I wrote the code, nor is it how it is being used throughout Cilium.

File: pkg/eventqueue/eventqueue.go [hunk 4]
```diff
@@ -167,13 +172,19 @@ func (ev *Event) WasCancelled() bool {
...
+// which one will return the channel which passes results back to the caller.
 func (q *EventQueue) Enqueue(ev *Event) <-chan interface{} {
-
 	if q.notSafeToAccess() || ev == nil {
 		return nil
 	}
 
+	// Events can only be enqueued once.
+	if atomic.AddInt32(&ev.enqueued, 1) > 1 {
+		return nil
```
Comment (aanm) [inline pkg/eventqueue/eventqueue.go]: This needs to be addressed somehow, it can be addressed in a different PR. If the caller of this function is **not** checking for the channel to be nil then doing `<-c` where `c` is a `nil` channel will deadlock forever.

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/eventqueue/eventqueue.go (hunk 3 above)
Comment (jrajahalme) [reply pkg/eventqueue/eventqueue.go:188]: To summarize: the `nil` return is an error condition the caller should check for.

File: pkg/eventqueue/eventqueue.go (hunk 4 above)
Comment (jrajahalme) [reply pkg/eventqueue/eventqueue.go]: Maybe document that the caller must check for `nil`, which is an error condition?

**Review by jrajahalme — APPROVED** [review]
Nit: document that the caller must check for error condition (`nil` return).

File: pkg/eventqueue/eventqueue.go (hunk 4 above)
Comment (jrajahalme) [reply pkg/eventqueue/eventqueue.go]: Note that `nil` return is done in other error conditions right above. Most likely all callers are already checking for `nil` return due to that.

Comment (aanm) [reply pkg/eventqueue/eventqueue.go]: > Note that `nil` return is done in other error conditions right above. Most likely all callers are already checking for `nil` return due to that.

@jrajahalme yeah the problem is that they are not. I've checked.

Comment (ianvernon) [reply pkg/eventqueue/eventqueue.go]: Fixed in all callpaths in latest commit. All call paths now check if the resultant channel is nil and return and / or log an appropriate error.

**Comment by ianvernon** [comment]
test-me-please

**Comment by ianvernon** [comment]
Build failure: https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Validated/14168/

* K8s NamespaceSelector policy test
* CNP Match Expressions test

Both failed due to healthchecks from the host.

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by ianvernon** [comment]
test-me-please

File: pkg/eventqueue/eventqueue.go (hunk 4 above)
Comment (joestringer) [reply pkg/eventqueue/eventqueue.go]: If this is an error case, then please return an error from the function. Golang has a ubiquitous standard for this, if there's an error, you return an error. Implementing functions that return error cases in a non-standard way is inviting future misuse.

**Review by joestringer — COMMENTED** [review]
What's the context of this? Is there actually a codepath where this is occurring or was it observed while experimenting with new uses of the eventqueue?

IMO we should return this as an error and have the callers actually check/handle it as an error case. It seems weirdly out of the expected model of the eventqueue for the same event to be queued multiple times.

**Comment by ianvernon** [comment]
@joestringer context is in https://github.com/cilium/cilium/issues/8805 . This was found while experimenting with the EventQueue. Happy to return an error, I'll change the function signature / associated callsites and re-upload.

**Comment by ianvernon** [comment]
test-me-please

File: pkg/eventqueue/eventqueue.go (hunk 4 above)
Comment (ianvernon) [reply pkg/eventqueue/eventqueue.go]: @joestringer fixed!

**Comment by ianvernon** [comment]
test-me-please
